src/streamlink/plugins/svtplay.py

```diff
@@ -1,92 +1,146 @@
+import logging
 import re
 
-from streamlink.plugin import Plugin
-from streamlink.plugin.api import StreamMapper, validate
-from streamlink.stream import HLSStream, HDSStream
+from streamlink.compat import urljoin
+from streamlink.exceptions import PluginError
+from streamlink.plugin import Plugin, PluginArguments, PluginArgument
+from streamlink.plugin.api import validate
+from streamlink.stream import DASHStream, HTTPStream
+from streamlink.stream.ffmpegmux import MuxedStream
 
-API_URL = "http://www.svt.se/videoplayer-api/video/{0}"
-
-_url_re = re.compile(r"""
-    http(s)?://
-    (www\.)?
-    (?:
-        svtplay |
-        svtflow |
-        oppetarkiv
-    )
-    .se
-""", re.VERBOSE)
-
-# Regex to match video ID
-_id_re = re.compile(r"""data-video-id=['"](?P<id>[^'"]+)['"]""")
-_old_id_re = re.compile(r"/(?:video|klipp)/(?P<id>[0-9]+)/")
-
-# New video schema used with API call
-_video_schema = validate.Schema(
-    {
-        "videoReferences": validate.all(
-            [{
-                "url": validate.text,
-                "format": validate.text
-            }],
-        ),
-    },
-    validate.get("videoReferences")
-)
-
-# Old video schema
-_old_video_schema = validate.Schema(
-    {
-        "video": {
-            "videoReferences": validate.all(
-                [{
-                    "url": validate.text,
-                    "playerType": validate.text
-                }],
-            ),
-        }
-    },
-    validate.get("video"),
-    validate.get("videoReferences")
-)
+log = logging.getLogger(__name__)
 
 
 class SVTPlay(Plugin):
+    api_url = 'https://api.svt.se/videoplayer-api/video/{0}'
+
+    author = None
+    category = None
+    title = None
+
+    url_re = re.compile(r'''
+        https?://(?:www\.)?(?:svtplay|oppetarkiv).se(?P<path>.*)
+    ''', re.VERBOSE)
+
+    latest_episode_url_re = re.compile(r'''
+        class="play_titlepage__latest-video"\s+href="(?P<url>[^"]+)"
+    ''', re.VERBOSE)
+
+    live_id_re = re.compile(r'.*/(?P<live_id>[^?]+)')
+
+    vod_id_re = re.compile(r'''
+        (?:DATA_LAKE\s+=\s+{"content":{"id":|data-video-id=)
+        "(?P<vod_id>[^"]+)"
+    ''', re.VERBOSE)
+
+    _video_schema = validate.Schema({
+        validate.optional('programTitle'): validate.text,
+        validate.optional('episodeTitle'): validate.text,
+        'videoReferences': [{
+            'url': validate.url(),
+            'format': validate.text,
+        }],
+        validate.optional('subtitleReferences'): [{
+            'url': validate.url(),
+            'format': validate.text,
+        }],
+    })
+
+    arguments = PluginArguments(
+        PluginArgument(
+            'mux-subtitles',
+            action='store_true',
+            help="Automatically mux available subtitles in to the output stream.",
+        ),
+    )
```

Comment:
Is it really necessary to add a plugin parameter for having muxed output streams with subtitles? We don't have this anywhere else, but if there is a need for having muxed subtitles, then maybe we could add a global parameter that could then be checked by individual plugins which have access to external subtitles. Having parameters for each plugin doesn't make much sense.

Reply:
Thanks for the review.

I only did it that way because I thought it was the 'right way to do it'!  It is the way it seems to have been done in the funimation, pluzz, rtve and vimeo plugins.

I added the subtitle support because I though it would be 'nice to have', but I have no great attachment to it either, so am happy to remove the support depending on how you want to proceed.

Cheers.

Reply:
> it seems to have been done in the funimation, pluzz, rtve and vimeo plugins

Apologies, this is indeed true.

We should change this though. It doesn't make sense for each plugin to have its own subtitle-muxing parameter. I hope you and anyone else agree with this.

I don't want to pressure you to change this now though. If you want to keep it, I guess that's fine considering we already have plugins with the same pattern. Deprecating or removing parameters however is always a bit tough. I don't want to decide this on my own.

Reply:
> We should change this though. It doesn't make sense for each plugin to have its own subtitle-muxing parameter. I hope you and anyone else will agree with this.

Seems perfectly reasonable to me.

> I don't want to pressure you to change this now though. If you want to keep it, I guess that's fine considering we already have plugins with the same pattern.

I'm happy either way.

> Deprecating or removing parameters however is always a bit tough. I don't want to decide this on my own.

It seems like an issue for you and the other main devs to discuss.  I guess if you decide to go the route of adding a global parameter, it will mean suppressing the ```--*-mux-subtitles``` entries in the above mentioned plugins for backwards compatibility and modifying them to use the new global version.

Reply:
I've been thinking about the implications of a global subtitle parameter.  Going that way also means asking why not do the same thing with other current plugin specific parameters which could be used by the plugin, or not, as necessary if they were global instead.  Likely candidates which are in no way absolutely plugin specific are email/username, password and purge-credentials, for example.

Reply:
> whereas a global parameter specified in the config file would apply to all plugins

not an issue, use this instead

https://streamlink.github.io/cli.html#plugin-specific-configuration-file

Reply:
Good point, I'd forgotten about the plugin specific config files.